### PR TITLE
Specify additional steps to utilize GPU for Linux users

### PR DESCRIPTION
**Pull request by sgkouzias**
Specify additional steps to utilize GPU for Linux users

### CODE REVIEW

**Comment by 8bitmp3** [comment]
@MarkDaoust @markmcd

**Review by sgkouzias — COMMENTED** [review]
Added steps and respective instructions to install TensorFlow by running the pip install tensorflow[and-cuda] command within a virtual environment (option #1: conda, option #2: venv) and set environment variables to find/locate compatible NVIDIA libs installed with TensorFlow to effectively utilize GPUs. The solution has been successfully tested.

Reference: https://github.com/tensorflow/tensorflow/issues/63362

**Comment by sgkouzias** [comment]
@haifeng-jin , @MarkDaoust, @8bitmp3  do we have any updates?

**Comment by sgkouzias** [comment]
> As I remembered, the current recommended way to install TF is to use `pip`. I do not have further info on this. @MarkDaoust may comment on this.

@haifeng-jin it seems practically impossible for someone owning a PC with CUDA-enabled GPU to perform deep learning experiments with TensorFlow version 2.16.1 and utilize his GPU locally without manually performing some extra steps not included (until today) in the official TensorFlow documentation of the standard installation procedure of TensorFlow for Linux users with GPUs at least as a temporal fix!

**Review by mihaimaruseac — COMMENTED** [review]
Please don't use "add file"/"update file"/"fix file"/etc. commit messages. These are hard to reason about when looking at the history of the file/repository. Instead, please write explanatory git commit messages.

The commit message is also the title of the PR if the PR has only one commit. It is thus twice important to have commit messages that are relevant, as PRs would be easier to understand and easier to analyze in search results.

For how to write good quality git commit messages, please consult https://cbea.ms/git-commit/

**Comment by mihaimaruseac** [comment]
> It turns out that when you `pip install tensorflow[and-cuda]` all required NVIDIA libraries are installed as well. You just need to configure manually the environment variables as appropriate in order to utilize them and run TensorFlow with GPU.

Can we instead add these to the install guide?

**Review by mihaimaruseac — APPROVED** [review]
I read the update and it seems reasonable to me. Thank you

**Comment by Tachi107** [comment]
Why is conda mentioned in this patch? It makes the install guide more convoluted and seems unnecessary to me.

**Comment by sgkouzias** [comment]
> Why is conda mentioned in this patch? It makes the install guide more convoluted and seems unnecessary to me.

@Tachi107 I agree. Should I proceed to erase everything related to conda refered as option 1 and just keep one suggested option (create a venv virtual environment)? Perhaps it would be better and more straight-forward?

**Comment by Tachi107** [comment]
Note that I'm not a tensorflow maintainer, just a casual user who happened to stumble upon this patch. But yeah, if I were you I would just show how to setup the venv. Conda users should already know how to do that with their non-default setup :)

**Comment by sgkouzias** [comment]
> Note that I'm not a tensorflow maintainer, just a casual user who happened to stumble upon this patch. But yeah, if I were you I would just show how to setup the venv. Conda users should already know how to do that with their non-default setup :)

@Tachi107 thank you. It seems very reasonable to simplify the guide like that. However for now I will keep it as is and await for the comments of the maintainers as well.

**Comment by sgkouzias** [comment]
@haifeng-jin , @MarkDaoust, @8bitmp3 I await any suggestions or revisions if needed. Do we have any updates?

**Comment by t-kalinowski** [comment]
There is no need to use conda, a standard venv works fine. In 2.15, tensorflow knew to go look for the NVIDIA binaries installed with `pip`. With TF 2.16, you can help it by placing them the binaries on LD_LIBRARY_PATH, like suggested in this PR, or by creating symlinks from the TF package to the pip installed nvidia packages. E.g., 

```bash
python -m venv my-venv
source my-venv/bin/activate
python -m pip install tensorflow[and-cuda]
pushd $(dirname $(python -c 'print(__import__("tensorflow").__file__)'))
ln -svf ../nvidia/*/lib/*.so* .
popd
```

This is essentially what we do from the R interface in `tensorflow::install_tensorflow()` and `keras3::install_keras()`

**Comment by sgkouzias** [comment]
@t-kalinowski thank you very much for your valuable advice. I revised the PR accordingly.

**Comment by t-kalinowski** [comment]
@sgkouzias if you also create a symlink at `my-venv/bin/ptxas` -> `my-venv/lib/python.../site-packages/.../bin/ptxax`, then you could probably get away without needing to require users to modify default `activate` and `deactivate` scripts.

**Comment by sgkouzias** [comment]
> @sgkouzias if you also create a symlink at `my-venv/bin/ptxas` -> `my-venv/lib/python.../site-packages/.../bin/ptxax`, then you could probably get away without needing to require users to modify default `activate` and `deactivate` scripts.

@t-kalinowski thank you so much for your advice. Instructions have been totally revised as per your comments. Modifications to default `activate` and `deactivate` scripts are not required from users. Instructions should resemble more or less what you do in the R interface.

File: site/en/install/pip.md
Comment (learning-to-play) [inline site/en/install/pip.md:210]: Can you remove `deactivate`?

Comment (learning-to-play) [reply site/en/install/pip.md]: I want to make sure that I understand the situation correctly. Which of the following two situation is correct?
- If the issue doesn't happen for 2.17.0RC0, yes please remove the instructions.
- If the issue happens for both 2.17.0RC0 and 2.16, we can wait for the GPU team to take a look at https://github.com/tensorflow/tensorflow/issues/63362 and see if the can send a fix for both 2.16.2 and 2.17.0 release.

Comment (sgkouzias) [reply site/en/install/pip.md]: The only difference is that on version 2.17.0.rc0 you only need the symlinks to nvidia libs in order to utilize GPUs while on version 2.16.1 you should in addition to creating symlinks to NVIDIA libs create symlink to ptxas. Eitherway the command `pip install tensorflow[and-cuda]` is not enough to utilize GPUs.

Comment (sgkouzias) [reply site/en/install/pip.md]: @learning-to-play unfortunately the latest release namely TensorFlow `2.16.2` does not fix the `ptxas` bug. When running a training script the user gets the error:

> ptxas returned an error during compilation of ptx to sass: 'INTERNAL: ptxas 12.3.103 has a bug that we think can affect XLA. Please use a different version.'  If the error message indicates that a file could not be written, please verify that sufficient filesystem space is provided.
> Aborted (core dumped)

**Comment by sgkouzias** [comment]
@learning-to-play, @SeeForTwo, @8bitmp3, @haifeng-jin, @MarkDaoust 

Unfortunately the latest release namely TensorFlow `2.16.2` does not fix the `ptxas` bug. When running a training script I get the error:

> ptxas returned an error during compilation of ptx to sass: 'INTERNAL: ptxas 12.3.103 has a bug that we think can affect XLA. Please use a different version.'  If the error message indicates that a file could not be written, please verify that sufficient filesystem space is provided.
> Aborted (core dumped)

So it seems as TensorFlow `2.16.2` Fails to work with GPUs as well !

**Notes:** 
1. Successful installation was verified by running:
`python3 -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"`
2. The solution included in the submitted pull request pending review helped to get rid of the `ptxas` bug and ultimately enforced TensorFlow `2.16.2` to work with my GPU:
```bash
ln -sf $(find $(dirname $(dirname $(python -c "import nvidia.cuda_nvcc; print(nvidia.cuda_nvcc.__file__)"))/*/bin/) -name ptxas -print -quit) $VIRTUAL_ENV/bin/ptxas
```

**Comment by belitskiy** [comment]
Thank you for the contribution, @sgkouzias :)
Given that the `[and-cuda]` installation now does detect pip-installed CUDA components again, please add a disclaimer specify that that symbolic links are only necessary in case the intended way doesn't work, i.e. the components aren't being detected, and/or conflict with the existing system CUDA installation (like `ptxas` for you).

**Comment by sgkouzias** [comment]
> Thank you for the contribution, @sgkouzias :) Given that the `[and-cuda]` installation now does detect pip-installed CUDA components again, please add a disclaimer specify that that symbolic links are only necessary in case the intended way doesn't work, i.e. the components aren't being detected, and/or conflict with the existing system CUDA installation (like `ptxas` for you).

@belitskiy, @learning-to-play I revised instructions as advised and will be awaiting your feedback. It is my honor to contribute to the TensorFlow community.

**Review by 8bitmp3 — DISMISSED** [review]
@MarkDaoust @markmcd PTAL

File: site/en/install/pip.md
Comment (MarkDaoust) [inline site/en/install/pip.md:238]:
```suggestion
    Check if TensorFlow successfully connected to the GPU
    
    ```
    import tensorflow as tf
    print("GPUs:", tf.config.list_physical_devices('GPU'))
    print(tf.constant(1) + tf.constant(1)) 
    ```

    If that doesn't work, you may need to add symbolic links because components aren't being
```

**Comment by MarkDaoust** [comment]
Thanks for all your work everyone (especially @sgkouzias)!

I just tweaked the order so that this new GPU debugging step is after the step where you test the GPU. 

I think this is still right so I'm merging it. But LMK if I misunderstood anything.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by sgkouzias** [comment]
> Thanks for all your work everyone (especially @sgkouzias)!
> 
> I just tweaked the order so that this new GPU debugging step is after the step where you test the GPU. 
> 
> I think this is still right so I'm merging it. But LMK if I misunderstood anything.

Thank you @MarkDaoust 🙏 it is my honour.
I noticed you mentioned merging, but it seems the pull request still needs a formal review due to branch protection rules. Could you please take a quick look and approve it when you have a chance?
Many thanks again!

**Review by MarkDaoust — APPROVED** [review]

**Comment by MarkDaoust** [comment]
Really it has everything it needs we're just waiting for the internal merge, it should be through soon.

**Pull request merged**
